nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -550,15 +563,21 @@ impl<'a> Planner<'a> {
             }
         }
 
-        // Double check that we didn't make any arithmetic mistakes. If we've
-        // arrived here, we think we've added the number of Nexus zones we
-        // needed to.
-        assert_eq!(
-            new_zones_added, num_zones_to_add,
-            "internal error counting {kind:?} zones"
-        );
-
-        Ok(())
+        // Double check that we didn't make any arithmetic mistakes.
+        // It's not an error to place fewer zones than requested;
+        // maybe they'll be added in the next round of planning.
```

Comment:
Relaxing this constraint seemed to me like the least invasive way to allow a kind of "soft failure" to allocate external DNS zones. The issue is timing: the `BuilderExternalNetworking` struct that tracks available DNS IP addresses is constructed before the planner gets to expunging zones in an expunged sled; therefore, we need to either do something like this, or make `BuilderExternalNetworking` more dynamic (which seems like a bad idea), or something like the previous filter, which I think we all agreed encoded more of the expunge logic than we would like.

Reply:
This seems reasonable to me, but I don't love that it's just a log message and won't be readily visible to an `omdb` user. Not a problem for this PR, certainly. I think we have other parts of the planner where we'd really like to user-visibly log why certain decisions were made; I'd add this to that pile.